### PR TITLE
Dataframe viewer scrolling

### PR DESCRIPTION
**Pull request by sheldonmlee**
I have polished up the scrolling implementation

### CODE REVIEW

**Comment by sheldonmlee** [comment]
Commit c37b74d5bea22d587a5c533271182c9c11fcda94 may need some reviewing to make sure it doesn't break stuff.

This commit was to address extra manual classification entries being created with the same ref, when going back and submitting previously submitted entries in 1083329e040c8787f605650daf665f96e03c57db.

File: classifier/rule_classifier.py [hunk 1]
```diff
@@ -151,14 +151,38 @@ def process_incoming_input(self, data):
         new_data = pd.DataFrame(data, columns=self.labeled_data.columns)
         new_data = new_data.astype(merged_types)
 
+        for i, tag in enumerate(new_data['Tags'].values):
```
Comment (EdwardBrodskiy) [inline classifier/rule_classifier.py]: Is this a fillna operation? If so please use pandas method

**Comment by EdwardBrodskiy** [comment]
Looking at what you did to process_incoming_input. Can't agree with the implementation. The original method was clean short and simple. It should not be touched. Modification should be moved to a separate method. This way you do not need to handle a bunch of edge cases. I would suggest i different implementation:

 - Each entered row has a edit and delete button which appear onHover
    - When a button is clicked this creates a popup with fields (would be good to reuse the ones from row entry if possible)
    - No index edge cases need to be handled as table index is implied by the button.
    - this then calls a new method in the rule classifier which can alter the internal data.

File: classifier/rule_classifier.py [hunk 2]
```diff
@@ -148,9 +148,11 @@ def process_user_input(data, part_labeled_row: pd.DataFrame):
         return user_inputs
 
     def process_incoming_input(self, data):
-        new_data = pd.DataFrame(data)
-        new_data.astype(merged_types)
+        new_data = pd.DataFrame(data, columns=self.labeled_data.columns)
```
Comment (EdwardBrodskiy) [inline classifier/rule_classifier.py]: This seems like a pointless addition as you use astype in the next line anyway

File: web_app/ingest_page/ingest_process.py [hunk 3]
```diff
@@ -30,16 +30,31 @@ def __init__(self, master, classifier, end_process=None, **kwargs):
...
+        self.last_entry_row = self.classifier.get_entry_prerequisites_for_manual_entry(self.interest_row)
+        self.row_entry.add_entry_row_at(*self.last_entry_row)
 
     def data_entered(self, data):
         if data is not None:
             self.classifier.process_incoming_input(data)
 
-        self.interest_row += 1
-        self.row_entry.add_entry_row_at(*self.classifier.get_entry_prerequisites_for_manual_entry(self.interest_row))
+        if self.reference_table.scroll(1):
```
Comment (EdwardBrodskiy) [inline web_app/ingest_page/ingest_process.py]: this if condition is unclear please specify with a comment or rename the function

File: web_app/ingest_page/ingest_process.py [hunk 4]
```diff
@@ -30,16 +30,31 @@ def __init__(self, master, classifier, end_process=None, **kwargs):
...
+        else:
+            self.row_entry.add_entry_row_at(*self.last_entry_row)
 
-        self.reference_table.scroll_down_one_row()
+        # move scroll to the end
+        self.update_idletasks()
+        self._parent_canvas.yview_moveto(1)
+
+    def go_back(self):
+        if self.reference_table.scroll(-1):
```
Comment (EdwardBrodskiy) [inline web_app/ingest_page/ingest_process.py]: same here

File: web_app/ingest_page/ingest_process.py (hunk 3 above)
Comment (EdwardBrodskiy) [reply web_app/ingest_page/ingest_process.py]: I am guessing this is handling the end of table case in which case you should consider re naming the method to "try_scroll_to" or "try_scroll"

File: web_app/ingest_page/row_entry.py [hunk 5]
```diff
@@ -8,13 +8,14 @@
 
 
 class RowEntry(customtkinter.CTkFrame):
-    def __init__(self, master, on_enter=None, **kwargs):
+    def __init__(self, master, on_enter=None,  **kwargs):
```
Comment (EdwardBrodskiy) [inline web_app/ingest_page/row_entry.py]: make sure you refactor your files before we merge to master

Comment (EdwardBrodskiy) [reply web_app/ingest_page/row_entry.py]: *double space

File: web_app/ingest_page/row_entry.py [hunk 6]
```diff
@@ -86,16 +85,41 @@ def submit(self):
...
+        self._container_config(container)
+        container.grid(row=self.row_index, column=0, columnspan=len(self.fields), sticky="ew", pady=5)
+        for i, row in enumerate(user_entries):
+            self.rows[ref]['label_rows'].append(self._draw_label_row(container, row, i))
             self.row_index += 1
 
         if self.on_enter is not None:
             self.on_enter(user_entries)
 
+    def _draw_label_row(self, parent, row, grid_row):
```
Comment (EdwardBrodskiy) [inline web_app/ingest_page/row_entry.py:99]: I am guessing all of this additional code is for that grouped row view in which that is fine. But maybe consider extracting it as its own component.

**Review by EdwardBrodskiy — CHANGES_REQUESTED** [review]

**Pull request merged**